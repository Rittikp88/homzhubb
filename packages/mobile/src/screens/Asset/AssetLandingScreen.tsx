import React from 'react';
import { StyleSheet, View, Image } from 'react-native';
import { withTranslation, WithTranslation } from 'react-i18next';
import { bindActionCreators, Dispatch } from 'redux';
import { connect } from 'react-redux';
import { LocaleConstants } from '@homzhub/common/src/services/Localization/constants';
import { StorageService, StorageKeys } from '@homzhub/common/src/services/storage/StorageService';
import { IRefreshTokenPayload, IUserPayload } from '@homzhub/common/src/domain/repositories/interfaces';
import { IState } from '@homzhub/common/src/modules/interfaces';
import { UserActions } from '@homzhub/common/src/modules/user/actions';
import { UserSelector } from '@homzhub/common/src/modules/user/selectors';
import { theme } from '@homzhub/common/src/styles/theme';
import { images } from '@homzhub/common/src/assets/images';
import { Text, Label, Button, SVGUri } from '@homzhub/common/src/components';
import { GradientBackground } from '@homzhub/mobile/src/components';
import { NavigationScreenProps, ScreensKeys } from '@homzhub/mobile/src/navigation/interfaces';
import { AppStackParamList } from '@homzhub/mobile/src/navigation/AppNavigator';
import { User } from '@homzhub/common/src/domain/models/User';

interface IStateProps {
  user: User | null;
}

interface IDispatchProps {
  logout: (data: IRefreshTokenPayload) => void;
}

type libraryProps = WithTranslation & NavigationScreenProps<AppStackParamList, ScreensKeys.PropertyPostLandingScreen>;
type Props = IStateProps & IDispatchProps & libraryProps;

export class AssetLandingScreen extends React.PureComponent<Props, {}> {
  public render(): React.ReactNode {
    const { t, user } = this.props;
    return (
      <GradientBackground>
        <View style={styles.container}>
          <View style={styles.imagesContainer}>
            <Image source={images.landingScreenLogo} />
            <Text type="regular" textType="semiBold">
              {t('welcomeUser', { username: user?.fullName })}
            </Text>
            <Text type="small" textType="regular" style={styles.description}>
              {t('description')}
            </Text>
            <SVGUri
              width={295}
              height={133}
              viewBox="0 0 295 133"
              preserveAspectRatio="xMidYMid meet"
              uri="https://homzhub-bucket.s3.amazonaws.com/Group+1168.svg"
            />
            <Label type="regular" textType="regular">
              {t('searchProject')}
            </Label>
          </View>
          <View style={styles.buttonContainer}>
            <Button
              title={t('addProperty')}
              type="primary"
              onPress={this.onAddProperty}
              containerStyle={styles.addProperty}
              testID="btnAddProperty"
            />
            {/* TODO: To be removed once the logic is in place */}
            <Button
              title="Dashboard"
              type="secondary"
              onPress={this.onSwitchProperty}
              containerStyle={styles.addProperty}
            />
            <Label type="large" textType="regular" style={styles.logoutContainer}>
              {t('logoutHelperText')} &nbsp;
              <Label type="large" textType="bold" style={styles.logout} onPress={this.logout} testID="lblLogout">
                {t('common:logout')}
              </Label>
            </Label>
          </View>
        </View>
      </GradientBackground>
    );
  }

  public onAddProperty = (): void => {
    const { navigation } = this.props;
    navigation.navigate(ScreensKeys.PropertyPostStack, { screen: ScreensKeys.AssetLocationSearch });
  };

  public onSwitchProperty = (): void => {
    const { navigation } = this.props;
    navigation.navigate(ScreensKeys.BottomTabs, { screen: ScreensKeys.Dashboard });
  };

  public logout = async (): Promise<void> => {
    const { logout } = this.props;
    const user: IUserPayload | null = (await StorageService.get(StorageKeys.USER)) ?? null;
    if (!user) {
      return;
    }
    const { refresh_token } = user;
    const logoutPayload = {
      refresh_token,
    };
    logout(logoutPayload);
  };
}

export const mapStateToProps = (state: IState): IStateProps => {
  return {
    user: UserSelector.getUserDetails(state),
  };
};

export const mapDispatchToProps = (dispatch: Dispatch): IDispatchProps => {
  const { logout } = UserActions;
  return bindActionCreators(
    {
      logout,
    },
    dispatch
  );
};

export default connect<IStateProps, IDispatchProps, WithTranslation, IState>(
  mapStateToProps,
  mapDispatchToProps
)(withTranslation(LocaleConstants.namespacesKey.property)(AssetLandingScreen));

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: theme.colors.white,
    shadowColor: theme.colors.shadow,
    shadowOffset: {
      width: 0,
      height: 10,
    },
    shadowOpacity: 0.51,
    shadowRadius: 13.16,
    elevation: 20,
    width: theme.viewport.width,
    maxHeight: theme.viewport.height - 100,
    borderRadius: 8,
    margin: theme.layout.screenPadding,
    paddingHorizontal: theme.layout.screenPadding,
  },
  imagesContainer: {
    justifyContent: 'space-around',
    alignItems: 'center',
    flex: 0.9,
  },
  description: {
    textAlign: 'center',
  },
  buttonContainer: {
    justifyContent: 'space-around',
    alignItems: 'stretch',
    marginVertical: 20,
  },
  addProperty: {
    flex: 0,
    marginVertical: 5,
  },
  logoutContainer: {
    alignSelf: 'center',
    marginVertical: 10,
  },
  logout: {
    color: theme.colors.blue,
  },
});
