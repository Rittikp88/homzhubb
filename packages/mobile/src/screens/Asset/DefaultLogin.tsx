import React from 'react';
import { View, StyleSheet } from 'react-native';
import { withTranslation, WithTranslation } from 'react-i18next';
import { CommonActions } from '@react-navigation/native';
import { theme } from '@homzhub/common/src/styles/theme';
import Logo from '@homzhub/common/src/assets/images/homzhubLogo.svg';
import { NavigationScreenProps, ScreensKeys } from '@homzhub/mobile/src/navigation/interfaces';
import { BottomTabNavigatorParamList } from '@homzhub/mobile/src/navigation/BottomTabs';
import { Button } from '@homzhub/common/src/components/atoms/Button';
import { Text } from '@homzhub/common/src/components/atoms/Text';

type libraryProps = WithTranslation & NavigationScreenProps<BottomTabNavigatorParamList, ScreensKeys.DefaultLogin>;
type IProps = libraryProps;

export class DefaultLogin extends React.PureComponent<IProps> {
  public render(): React.ReactElement {
    const { t } = this.props;
    return (
      <View style={styles.container}>
        <Logo height={100} width={100} />
        <Text type="small" textType="regular" style={styles.loginText}>
          {t('pleaseSignup')}
        </Text>
        <Button type="primary" title={t('signUp')} containerStyle={styles.button} onPress={this.navigateToSignup} />
      </View>
    );
  }

  public navigateToSignup = (): void => {
    const { navigation } = this.props;
    navigation.dispatch(
      CommonActions.navigate({
        name: ScreensKeys.AuthStack,
      })
    );
  };
}

export default withTranslation()(DefaultLogin);

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: theme.colors.white,
  },
  loginText: {
    marginVertical: 10,
  },
  button: {
    flex: 0,
    marginVertical: 10,
  },
});
