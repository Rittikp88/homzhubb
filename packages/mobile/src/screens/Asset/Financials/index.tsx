import React from 'react';
import { StyleSheet } from 'react-native';
import { withTranslation, WithTranslation } from 'react-i18next';
import { AlertHelper } from '@homzhub/mobile/src/utils/AlertHelper';
import { ErrorUtils } from '@homzhub/common/src/utils/ErrorUtils';
import { DateUtils } from '@homzhub/common/src/utils/DateUtils';
import { LedgerUtils } from '@homzhub/common/src/utils/LedgerUtils';
import { LedgerService } from '@homzhub/common/src/services/LedgerService';
import { LocaleConstants } from '@homzhub/common/src/services/Localization/constants';
import { FinancialsNavigatorParamList } from '@homzhub/mobile/src/navigation/BottomTabs';
import { NavigationScreenProps, ScreensKeys } from '@homzhub/mobile/src/navigation/interfaces';
import {
  AnimatedProfileHeader,
  FinancialHeaderContainer,
  PropertyDuesCardContainer,
  StateAwareComponent,
} from '@homzhub/mobile/src/components';
import FinanceOverview from '@homzhub/mobile/src/components/organisms/FinanceOverview';
import TransactionCardsContainer from '@homzhub/mobile/src/components/organisms/TransactionCardsContainer';
import { propertyDues } from '@homzhub/common/src/mocks/FinancialsTabMockData';
import { DataGroupBy, GeneralLedgers, LedgerTypes } from '@homzhub/common/src/domain/models/GeneralLedgers';
import { FinancialRecords, FinancialTransactions } from '@homzhub/common/src/domain/models/FinancialTransactions';

interface IState {
  ledgerData: GeneralLedgers[];
  transactionsData: FinancialRecords[];
  isLoading: boolean;
  scrollEnabled: boolean;
}

type Props = WithTranslation & NavigationScreenProps<FinancialsNavigatorParamList, ScreensKeys.FinancialsLandingScreen>;

export class Financials extends React.PureComponent<Props, IState> {
  private onFocusSubscription: any;

  public state = {
    ledgerData: [],
    transactionsData: [],
    isLoading: false,
    scrollEnabled: true,
  };

  public componentDidMount(): void {
    const { navigation } = this.props;

    this.onFocusSubscription = navigation.addListener('focus', (): void => {
      this.getGeneralLedgersPref().then();
      this.getGeneralLedgers().then();
    });
  }

  public render = (): React.ReactElement => {
    const { isLoading } = this.state;
    return <StateAwareComponent loading={isLoading} renderComponent={this.renderComponent()} />;
  };

  private renderComponent = (): React.ReactElement => {
    const { t } = this.props;
    const { ledgerData, transactionsData, scrollEnabled } = this.state;
    const { currency_symbol, totalDue, details } = propertyDues;

    return (
      <AnimatedProfileHeader isOuterScrollEnabled={scrollEnabled} title={t('financial')}>
        <>
          <FinancialHeaderContainer
            onPlusIconClicked={this.onPlusIconPress}
            // @ts-ignore
            title={t('assetFinancial:recordsText')}
            income={LedgerUtils.getSumOfTransactionsOfType(LedgerTypes.credit, ledgerData)}
            expense={LedgerUtils.getSumOfTransactionsOfType(LedgerTypes.debit, ledgerData)}
            individualCardStyle={styles.individualCardStyle}
          />
          <FinanceOverview />
          <PropertyDuesCardContainer currency={currency_symbol} totalDue={totalDue} propertyDues={details} />
          {transactionsData && transactionsData.length > 0 && (
            <TransactionCardsContainer
              shouldEnableOuterScroll={this.toggleScroll}
              transactionsData={transactionsData}
              onEndReachedHandler={this.onEndReachedHandler}
            />
          )}
        </>
      </AnimatedProfileHeader>
    );
  };

  private onEndReachedHandler = async (): Promise<void> => {
    await this.getGeneralLedgers();
  };

  private onPlusIconPress = (): void => {
    const {
      navigation: { navigate },
    } = this.props;
    navigate(ScreensKeys.AddRecordScreen);
  };

  private toggleScroll = (scrollEnabled: boolean): void => {
    this.setState({ scrollEnabled });
  };

  private getGeneralLedgersPref = async (): Promise<void> => {
    this.setState({ isLoading: true });
    try {
      const response: GeneralLedgers[] = await LedgerService.getLedgerPerformances(
        DateUtils.getCurrentYearStartDate(),
        DateUtils.getCurrentYearLastDate(),
        DataGroupBy.year
      );

      this.setState({ ledgerData: response, isLoading: false });
    } catch (e) {
      this.setState({ isLoading: false });
      AlertHelper.error({ message: ErrorUtils.getErrorMessage(e.details) });
    }
  };

  private getGeneralLedgers = async (): Promise<void> => {
    const { transactionsData } = this.state;
    this.setState(() => {
      return { isLoading: false };
    });

    try {
      const response: FinancialTransactions = await LedgerService.getGeneralLedgers(transactionsData.length);

      this.setState({
        transactionsData: [...transactionsData, ...response.results],
        isLoading: false,
      });
    } catch (e) {
      this.setState({ isLoading: false });
      AlertHelper.error({ message: ErrorUtils.getErrorMessage(e.details) });
    }
  };
}

export default withTranslation(LocaleConstants.namespacesKey.assetFinancial)(Financials);

const styles = StyleSheet.create({
  individualCardStyle: {
    alignItems: 'center',
    paddingHorizontal: 25,
  },
});
