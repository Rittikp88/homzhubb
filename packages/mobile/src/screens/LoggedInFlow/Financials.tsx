import React from 'react';
import { View, StyleSheet, Text } from 'react-native';

export class Financials extends React.PureComponent<{}, {}> {
  public render = (): React.ReactElement => {
    return (
      <View style={styles.screen}>
        <Text>Financials Screen</Text>
      </View>
    );
  };
}

const styles = StyleSheet.create({
  screen: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
});
