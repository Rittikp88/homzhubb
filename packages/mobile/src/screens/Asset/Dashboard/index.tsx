import React from 'react';
import { FlatList, TouchableOpacity, ViewStyle, TextStyle, StyleProp } from 'react-native';
import { withTranslation, WithTranslation } from 'react-i18next';
import { bindActionCreators, Dispatch } from 'redux';
import { connect } from 'react-redux';
import { AlertHelper } from '@homzhub/common/src/utils/AlertHelper';
import { AnalyticsService } from '@homzhub/common/src/services/Analytics/AnalyticsService';
import { ErrorUtils } from '@homzhub/common/src/utils/ErrorUtils';
import { EventType } from '@homzhub/common/src/services/Analytics/EventType';
import { AssetRepository } from '@homzhub/common/src/domain/repositories/AssetRepository';
import { DashboardRepository } from '@homzhub/common/src/domain/repositories/DashboardRepository';
import { PortfolioActions } from '@homzhub/common/src/modules/portfolio/actions';
import { RecordAssetActions } from '@homzhub/common/src/modules/recordAsset/actions';
import { UserActions } from '@homzhub/common/src/modules/user/actions';
import { UserSelector } from '@homzhub/common/src/modules/user/selectors';
import { theme } from '@homzhub/common/src/styles/theme';
import Icon, { icons } from '@homzhub/common/src/assets/icon';
import { AssetAdvertisementBanner, AssetMetricsList, AssetSummary } from '@homzhub/mobile/src/components';
import { Text } from '@homzhub/common/src/components/atoms/Text';
import { BottomSheet } from '@homzhub/common/src/components/molecules/BottomSheet';
import AssetMarketTrends from '@homzhub/mobile/src/components/molecules/AssetMarketTrends';
import UserSubscriptionPlan from '@homzhub/common/src/components/molecules/UserSubscriptionPlan';
import FinanceOverview from '@homzhub/mobile/src/components/organisms/FinanceOverview';
import PendingPropertyListCard from '@homzhub/mobile/src/components/organisms/PendingPropertyListCard';
import { UserScreen } from '@homzhub/mobile/src/components/HOC/UserScreen';
import { Asset, PropertyStatus } from '@homzhub/common/src/domain/models/Asset';
import { AssetMetrics } from '@homzhub/common/src/domain/models/AssetMetrics';
import { Filters } from '@homzhub/common/src/domain/models/AssetFilter';
import { IActions, ISelectedAssetPlan } from '@homzhub/common/src/domain/models/AssetPlan';
import { LocaleConstants } from '@homzhub/common/src/services/Localization/constants';
import { NavigationScreenProps, ScreensKeys } from '@homzhub/mobile/src/navigation/interfaces';
import { DashboardNavigatorParamList } from '@homzhub/mobile/src/navigation/DashboardStack';
import { IState } from '@homzhub/common/src/modules/interfaces';
import { ISetAssetPayload } from '@homzhub/common/src/modules/portfolio/interfaces';
import { IApiClientError } from '@homzhub/common/src/network/ApiClientError';

interface IDispatchProps {
  setCurrentFilter: (payload: Filters) => void;
  setAssetId: (payload: number) => void;
  setSelectedPlan: (payload: ISelectedAssetPlan) => void;
  setAddPropertyFlow: (payload: boolean) => void;
  setCurrentAsset: (payload: ISetAssetPayload) => void;
  setInitialState: () => void;
}

interface IReduxStateProps {
  assets: Asset[];
}

type libraryProps = NavigationScreenProps<DashboardNavigatorParamList, ScreensKeys.DashboardLandingScreen>;
type Props = WithTranslation & libraryProps & IDispatchProps & IReduxStateProps;

interface IDashboardState {
  metrics: AssetMetrics;
  pendingProperties: Asset[];
  isLoading: boolean;
  showBottomSheet: boolean;
}

interface IFormattedBottomSheetData {
  icon: string;
  label: string;
  onPress: () => void;
}

const ShowInMvpRelease = false;

export class Dashboard extends React.PureComponent<Props, IDashboardState> {
  public focusListener: any;

  public state = {
    metrics: {} as AssetMetrics,
    pendingProperties: [],
    isLoading: false,
    showBottomSheet: false,
  };

  public componentDidMount = (): void => {
    const { navigation } = this.props;
    this.focusListener = navigation.addListener('focus', () => {
      this.getScreenData().then();
    });
  };

  public componentWillUnmount(): void {
    this.focusListener();
  }

  public render = (): React.ReactElement => {
    const { t } = this.props;
    const { isLoading, pendingProperties } = this.state;

    return (
      <UserScreen loading={isLoading} isGradient title={t('dashboard')}>
        {this.renderAssetMetricsAndUpdates()}
        {pendingProperties.length > 0 && (
          <PendingPropertyListCard
            data={pendingProperties}
            onPressComplete={this.onCompleteDetails}
            onSelectAction={this.handleActionSelection}
            onViewProperty={this.onViewProperty}
          />
        )}
        <FinanceOverview />
        <AssetMarketTrends isDashboard onViewAll={this.onViewAll} onTrendPress={this.onTrendPress} />
        <AssetAdvertisementBanner />
        {ShowInMvpRelease && <UserSubscriptionPlan onApiFailure={this.onAssetSubscriptionApiFailure} />}
        {this.renderBottomSheet()}
      </UserScreen>
    );
  };

  public renderAssetMetricsAndUpdates = (): React.ReactElement => {
    const { metrics } = this.state;
    return (
      <>
        <AssetMetricsList
          title={`${metrics?.assetMetrics?.assets?.count ?? 0}`}
          data={metrics?.assetMetrics?.miscellaneous ?? []}
          subscription={metrics?.userServicePlan?.label}
          onMetricsClicked={this.handleMetricsNavigation}
          onPlusIconClicked={this.openBottomSheet}
        />
        <AssetSummary
          notification={metrics?.updates?.notifications?.count ?? 0}
          serviceTickets={metrics?.updates?.tickets?.open?.count ?? 0}
          dues={metrics?.updates?.dues?.count ?? 0}
          messages={metrics?.updates?.messages?.unread?.count ?? 0}
          containerStyle={styles.assetCards()}
          onPressDue={this.handleDues}
          onPressServiceTickets={this.handleServiceTickets}
          onPressNotification={this.handleNotification}
          onPressMessages={this.handleMessages}
        />
      </>
    );
  };

  public renderBottomSheet = (): React.ReactElement => {
    const { t, setInitialState } = this.props;
    const { showBottomSheet } = this.state;
    const data = this.formatBottomSheetData();

    const keyExtractor = (item: IFormattedBottomSheetData, index: number): string => `${item}:${index}`;

    const renderItem = ({ item, index }: { item: IFormattedBottomSheetData; index: number }): React.ReactElement => {
      const { icon, label, onPress } = item;

      const onPressItem = (): void => {
        onPress();
        setInitialState();
        this.closeBottomSheet();
      };
      return (
        <>
          <TouchableOpacity
            style={[styles.bottomSheetItemContainer(), index % 2 === 0 && styles.evenItem()]}
            onPress={onPressItem}
          >
            <Icon name={icon} size={25} color={theme.colors.blue} />
            <Text type="small" textType="regular" style={styles.itemLabel()}>
              {label}
            </Text>
          </TouchableOpacity>
        </>
      );
    };

    return (
      <BottomSheet
        headerTitle={t('common:addCamelCase')}
        visible={showBottomSheet}
        onCloseSheet={this.closeBottomSheet}
        sheetHeight={375}
      >
        <FlatList
          data={data}
          numColumns={2}
          keyExtractor={keyExtractor}
          renderItem={renderItem}
          contentContainerStyle={styles.flatList()}
        />
      </BottomSheet>
    );
  };

  // HANDLERS
  private onAssetSubscriptionApiFailure = (err: IApiClientError): void => {
    AlertHelper.error({ message: ErrorUtils.getErrorMessage(err) });
  };

  private onCompleteDetails = (assetId: number): void => {
    const { navigation, setAssetId } = this.props;
    setAssetId(assetId);
    // @ts-ignore
    navigation.navigate(ScreensKeys.PropertyPostStack, {
      screen: ScreensKeys.AddProperty,
      params: { previousScreen: ScreensKeys.Dashboard },
    });
  };

  private onViewAll = (): void => {
    const { navigation } = this.props;
    navigation.navigate(ScreensKeys.MarketTrends, { isFromDashboard: true });
  };

  private onTrendPress = (url: string, trendId: number): void => {
    const { navigation } = this.props;
    navigation.navigate(ScreensKeys.WebViewScreen, { url, trendId });
  };

  private onViewProperty = (data: ISetAssetPayload): void => {
    const { setCurrentAsset, navigation } = this.props;
    setCurrentAsset(data);
    navigation.navigate(ScreensKeys.PropertyDetailScreen);
  };

  private handleMessages = (): void => {
    const { navigation } = this.props;
    navigation.navigate(ScreensKeys.Messages, { isFromDashboard: true });
  };

  private handleDues = (): void => {
    /**
     *
     navigation.dispatch(
     CommonActions.reset({
        index: 0,
        routes: [{ name: ScreensKeys.Financials }],
      })
     );
     */
    const { navigation, t } = this.props;
    navigation.navigate(ScreensKeys.ComingSoonScreen, { title: t('dues') ?? '', tabHeader: t('dashboard') });
  };

  private handleServiceTickets = (): void => {
    const { navigation } = this.props;
    navigation.navigate(ScreensKeys.ServiceTicketScreen, { isFromDashboard: true });
  };

  private handleNotification = (): void => {
    const { navigation } = this.props;
    navigation.navigate(ScreensKeys.AssetNotifications, { isFromDashboard: true });
  };

  private handleMetricsNavigation = (name: string): void => {
    const { navigation, setCurrentFilter } = this.props;
    setCurrentFilter(name as Filters);
    // @ts-ignore
    navigation.navigate(ScreensKeys.Portfolio, {
      screen: ScreensKeys.PortfolioLandingScreen,
    });
  };

  private handleActionSelection = (item: IActions, assetId: number): void => {
    const { navigation, setSelectedPlan, setAssetId } = this.props;
    setSelectedPlan({ id: item.id, selectedPlan: item.type });
    setAssetId(assetId);
    // @ts-ignore
    navigation.navigate(ScreensKeys.PropertyPostStack, {
      screen: ScreensKeys.AssetListing,
      params: { previousScreen: ScreensKeys.Dashboard },
    });
  };

  private openBottomSheet = (): void => this.setState({ showBottomSheet: true });

  private closeBottomSheet = (): void => this.setState({ showBottomSheet: false });

  private formatBottomSheetData = (): IFormattedBottomSheetData[] => {
    const { t, navigation, assets } = this.props;

    const handleAddProperty = (): void => {
      // @ts-ignore
      navigation.navigate(ScreensKeys.PropertyPostStack, { screen: ScreensKeys.AssetLocationSearch });
      AnalyticsService.track(EventType.AddPropertyInitiation);
    };

    const handleAddFinancialRecord = (): void => {
      if (assets.length <= 0) {
        AlertHelper.error({ message: t('assetFinancial:addProperty') });
        return;
      }
      navigation.navigate(ScreensKeys.AddRecordScreen, { isFromDashboard: true });
    };

    const handleAddServiceTicket = (): void => {
      if (assets.length <= 0) {
        AlertHelper.error({ message: t('assetDashboard:addPropertyForTicket') });
        return;
      }
      navigation.navigate(ScreensKeys.AddServiceTicket, { isFromDashboard: true });
    };

    const handleAddSupportTicket = (): void => {
      navigation.navigate(ScreensKeys.SupportScreen, { isFromDashboard: true });
    };

    const formattedDetails = [
      {
        icon: icons.portfolioFilled,
        label: t('assetFinancial:property'),
        onPress: handleAddProperty,
      },
      {
        icon: icons.barChartFilled,
        label: t('assetDashboard:incomeOrExpense'),
        onPress: handleAddFinancialRecord,
      },
      {
        icon: icons.serviceTicket,
        label: t('assetDashboard:serviceTicket'),
        onPress: handleAddServiceTicket,
      },
      {
        icon: icons.supportTicket,
        label: t('assetDashboard:supportTicket'),
        onPress: handleAddSupportTicket,
      },
    ];
    return formattedDetails;
  };

  // HANDLERS end

  // APIs
  private getScreenData = async (): Promise<void> => {
    await this.getAssetMetrics();
    await this.getPendingProperties();
  };

  private getAssetMetrics = async (): Promise<void> => {
    this.setState({ isLoading: true });
    try {
      const response: AssetMetrics = await DashboardRepository.getAssetMetrics('v3');
      this.setState({ metrics: response, isLoading: false });
    } catch (e) {
      this.setState({ isLoading: false });
      const error = ErrorUtils.getErrorMessage(e.details);
      AlertHelper.error({ message: error });
    }
  };

  private getPendingProperties = async (): Promise<void> => {
    try {
      const response: Asset[] = await AssetRepository.getPropertiesByStatus(PropertyStatus.PENDING);
      this.setState({ pendingProperties: response });
    } catch (e) {
      AlertHelper.error({ message: ErrorUtils.getErrorMessage(e.details) });
    }
  };
  // APIs end
}

export const mapDispatchToProps = (dispatch: Dispatch): IDispatchProps => {
  const { setCurrentFilter, setCurrentAsset, setInitialState } = PortfolioActions;
  const { setAddPropertyFlow } = UserActions;
  const { setAssetId, setSelectedPlan } = RecordAssetActions;
  return bindActionCreators(
    { setCurrentFilter, setAssetId, setSelectedPlan, setAddPropertyFlow, setCurrentAsset, setInitialState },
    dispatch
  );
};

const mapStateToProps = (state: IState): IReduxStateProps => {
  const { getUserAssets } = UserSelector;
  return {
    assets: getUserAssets(state),
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withTranslation(LocaleConstants.namespacesKey.assetDashboard)(Dashboard));

const styles = {
  assetCards: (): StyleProp<ViewStyle> => ({
    marginVertical: 12,
  }),
  evenItem: (): StyleProp<ViewStyle> => ({
    marginEnd: 17,
  }),
  bottomSheetItemContainer: (): StyleProp<ViewStyle> => ({
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    paddingHorizontal: 26,
    paddingVertical: 20,
    marginBottom: 20,
    borderWidth: 1,
    borderColor: theme.colors.disabled,
    borderRadius: 4,
  }),
  itemLabel: (): StyleProp<TextStyle> => ({
    marginTop: 10,
    textAlign: 'center',
    color: theme.colors.blue,
  }),
  flatList: (): StyleProp<ViewStyle> => ({
    marginBottom: 30,
    marginHorizontal: 24,
  }),
};
