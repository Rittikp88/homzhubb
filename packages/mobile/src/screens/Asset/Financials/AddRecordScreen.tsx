import React, { ReactElement } from 'react';
import { WithTranslation, withTranslation } from 'react-i18next';
import { StyleSheet } from 'react-native';
import { connect } from 'react-redux';
import { AlertHelper } from '@homzhub/mobile/src/utils/AlertHelper';
import { LedgerRepository } from '@homzhub/common/src/domain/repositories/LedgerRepository';
import { UserSelector } from '@homzhub/common/src/modules/user/selectors';
import { LocaleConstants } from '@homzhub/common/src/services/Localization/constants';
import { NavigationScreenProps, ScreensKeys } from '@homzhub/mobile/src/navigation/interfaces';
import { FinancialsNavigatorParamList } from '@homzhub/mobile/src/navigation/BottomTabs';
import { AnimatedProfileHeader, HeaderCard, StateAwareComponent } from '@homzhub/mobile/src/components';
import AddRecordForm from '@homzhub/mobile/src/components/organisms/AddRecordForm';
import { Asset } from '@homzhub/common/src/domain/models/Asset';
import { Currency } from '@homzhub/common/src/domain/models/Currency';
import { LedgerCategory } from '@homzhub/common/src/domain/models/LedgerCategory';
import { IState } from '@homzhub/common/src/modules/interfaces';

interface IScreenState {
  ledgerCategories: LedgerCategory[];
  clearForm: boolean;
  isLoading: boolean;
}

interface IStateToProps {
  currency: Currency;
  assets: Asset[];
}

type libraryProps = WithTranslation & NavigationScreenProps<FinancialsNavigatorParamList, ScreensKeys.AddRecordScreen>;
type IProps = libraryProps & IStateToProps;

export class AddRecordScreen extends React.PureComponent<IProps, IScreenState> {
  public state = {
    ledgerCategories: [],
    isLoading: false,
    clearForm: false,
  };

  public async componentDidMount(): Promise<void> {
    this.setState({ isLoading: true });

    const categories = await LedgerRepository.getLedgerCategories();

    this.setState({ ledgerCategories: categories, isLoading: false });
  }

  public render(): ReactElement {
    const { isLoading } = this.state;

    return <StateAwareComponent loading={isLoading} renderComponent={this.renderComponent()} />;
  }

  private renderComponent = (): ReactElement => {
    const { t } = this.props;
    const { clearForm } = this.state;
    return (
      <AnimatedProfileHeader title={t('financial')}>
        <>
          <HeaderCard
            title={t('addRecords')}
            clear={clearForm}
            subTitle={t('common:clear')}
            renderItem={this.renderAddRecordForm}
            onIconPress={this.goBack}
            onClearPress={this.onClearPress}
          />
        </>
      </AnimatedProfileHeader>
    );
  };

  private renderAddRecordForm = (): ReactElement => {
    const { clearForm, ledgerCategories } = this.state;
    const { currency, assets } = this.props;

    return (
      <AddRecordForm
        properties={assets}
        ledgerCategories={ledgerCategories}
        clear={clearForm}
        defaultCurrency={currency}
        onFormClear={this.onClearPress}
        containerStyles={styles.addFormContainer}
        onSubmitFormSuccess={this.onSubmitFormSuccess}
        shouldLoad={this.toggleLoading}
      />
    );
  };

  private onClearPress = (): void => {
    const { clearForm } = this.state;
    this.setState({ clearForm: !clearForm });
  };

  private onSubmitFormSuccess = (): void => {
    const { t } = this.props;
    AlertHelper.success({ message: t('addedSuccessfullyMessage') });
    this.goBack();
  };

  private toggleLoading = (isLoading: boolean): void => {
    this.setState({ isLoading });
  };

  private goBack = (): void => {
    const { navigation } = this.props;

    if (navigation.canGoBack()) {
      navigation.goBack();
    }
  };
}

const mapStateToProps = (state: IState): IStateToProps => {
  return {
    currency: UserSelector.getCurrency(state),
    assets: UserSelector.getUserAssets(state),
  };
};

export default connect(
  mapStateToProps,
  null
)(withTranslation(LocaleConstants.namespacesKey.assetFinancial)(AddRecordScreen));

const styles = StyleSheet.create({
  addFormContainer: {
    marginTop: 24,
  },
});
