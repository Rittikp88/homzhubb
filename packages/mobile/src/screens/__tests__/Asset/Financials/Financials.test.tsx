// @ts-nocheck
import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Financials } from '@homzhub/mobile/src/screens/Asset/Financials';

describe('Financials Screen', () => {
  let component: ShallowWrapper;

  beforeEach(() => {
    component = shallow(<Financials t={(key: string): string => key} />);
  });

  it('should render snapshot', () => {
    expect(toJson(component)).toMatchSnapshot();
  });
});
