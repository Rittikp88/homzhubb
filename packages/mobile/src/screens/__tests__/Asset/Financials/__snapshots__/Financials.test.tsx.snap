// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Financials Screen should render snapshot 1`] = `
<Fragment>
  <Memo(AnimatedProfileHeader)
    isOuterScrollEnabled={true}
    title="financial"
  >
    <Memo(AssetMetricsList)
      data={
        Array [
          Object {
            "colorGradient": Object {
              "hexColorA": "#61D773",
              "hexColorB": "#94EF90",
              "location": Array [
                0,
                1,
              ],
            },
            "count": "0/-",
            "currencySymbol": "INR",
            "label": "assetFinancial:income",
          },
          Object {
            "colorGradient": Object {
              "hexColorA": "#FDB113",
              "hexColorB": "#FFDB8F",
              "location": Array [
                0,
                1,
              ],
            },
            "count": "0/-",
            "currencySymbol": "INR",
            "label": "assetFinancial:expense",
          },
        ]
      }
      isSubTextRequired={false}
      numOfElements={2}
      onPlusIconClicked={[Function]}
      textStyle={
        Object {
          "color": "#FFFFFF",
          "textAlign": "center",
        }
      }
      title="assetFinancial:recordsText"
    />
    <Connect(withI18nextTranslation(FinanceOverview)) />
  </Memo(AnimatedProfileHeader)>
  <Loader
    visible={false}
  />
</Fragment>
`;
